### PR TITLE
Write Read, Replace and Patch ReplicaSetScale test +3 endpoints 

### LINKED ISSUE #98920: Write Read, Replace and Patch ReplicaSetScale test +3 endpoints
# Progress <code>[5/6]</code>

-   [X] APISnoop org-flow : [ReplicaSetScaleTest.org](https://github.com/apisnoop/ticket-writing/blob/AppsEndpoints/ReplicaSetScaleMockTest.org)
-   [X] test approval issue : [kubernetes/kubernetes#98920](https://github.com/kubernetes/kubernetes/issues/98920)
-   [x] test pr : kuberenetes/kubernetes[#98987](https://github.com/kubernetes/kubernetes/pull/98987)
-   [X] two weeks soak start date : [testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=Replicaset%20should%20have%20a%20working%20scale%20subresource)
-   [ ] two weeks soak end date : 26 Feb 2021
-   [x] test promotion pr : kubernetes/kubernetes[#99282](https://github.com/kubernetes/kubernetes/pull/99282)

# Identifying an untested feature Using APISnoop

According to this APIsnoop query, there are still some remaining RESOURCENAME endpoints which are untested.

with this query you can filter untested endpoints by their category and eligiblity for conformance. e.g below shows a query to find all conformance eligible untested,stable,core endpoints

```sql-mode
SELECT
  endpoint,
  -- k8s_action,
  -- path,
  -- description,
  kind
  FROM testing.untested_stable_endpoint
  where eligible is true
  and endpoint like '%ReplicaSetScale%'
  --and category = 'core'
  order by kind, endpoint desc
  limit 25;
```

```example
                endpoint                | kind
----------------------------------------|-------
 replaceAppsV1NamespacedReplicaSetScale | Scale
 readAppsV1NamespacedReplicaSetScale    | Scale
 patchAppsV1NamespacedReplicaSetScale   | Scale
(3 rows)

```

# API Reference and feature documentation

-   [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#read-scale-replicaset-v1-apps)
-   [client-go - ReplicaSet](https://pkg.go.dev/k8s.io/kubernetes/test/e2e/framework/replicaset)

# The mock test

## Test outline

1.  Create a ReplicaSet with a static label

2.  Read the ReplicaSetScale

3.  Update the ReplicaSetScale

4.  Get the ReplicaSetScale to ensure it is updated

5.  Patch the ReplicaSetScale

6.  Read the ReplicaSetScale to ensure it is Patch

7.  Delete the Namespace and ReplicaSet

### Test the functionality in Go

Due to the complexity of setting up the resources for APIService we have used the current e2e test as template. It has been extended in a [new ginkgo test](https://github.com/ii/kubernetes/commit/a157574c27e0efa55d3ce2b742fb45bb451a1ecf#diff-c4323b71233a6ca89375ef859d5757a50af9bbeb842d279cdd8f7f80fc0fd882R352-R420) for review.

# Verifying increase in coverage with APISnoop

```sql-mode
select distinct  endpoint, useragent
                 -- to_char(to_timestamp(release_date::bigint), ' HH:MI') as time
from testing.audit_event
where endpoint ilike '%ReplicaSetScale%'
  -- and release_date::BIGINT > round(((EXTRACT(EPOCH FROM NOW()))::numeric)*1000,0) - 60000
-- and useragent like 'kubectl%'
order by endpoint
limit 100;

```

```example
                endpoint                |                                                          useragent
----------------------------------------|------------------------------------------------------------------------------------------------------------------------------
 patchAppsV1NamespacedReplicaSetScale   | e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-apps] ReplicaSet Replicaset should have a working scale subresource
 readAppsV1NamespacedReplicaSetScale    | e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-apps] ReplicaSet Replicaset should have a working scale subresource
 replaceAppsV1NamespacedReplicaSetScale | e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format -- [sig-apps] ReplicaSet Replicaset should have a working scale subresource
(3 rows)

```

# Final notes

If a test with these calls gets merged, ****test coverage will go up by 3 points****

This test is also created with the goal of conformance promotion.

---

/sig testing
/sig architecture

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
- replaceAppsV1NamespacedReplicaSetScale
- readAppsV1NamespacedReplicaSetScale
- patchAppsV1NamespacedReplicaSetScale

**Which issue(s) this PR fixes:**
Fixes #98920

**Testgrid Link:**
[Testgrid](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=Replicaset%20should%20have%20a%20working%20scale%20subresource)

**Special notes for your reviewer:**
Adds +3 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/sig apps
/area conformance

### CODE REVIEW

**Comment by Riaankl** [comment]
/test pull-kubernetes-e2e-kind

Unrelated flake
```
• Failure [146.959 seconds]
[sig-storage] PersistentVolumes-local 
test/e2e/storage/utils/framework.go:23
  [Volume type: dir-link-bindmounted]
  test/e2e/storage/persistent_volumes-local.go:191
    Two pods mounting a local volume one after the other
    test/e2e/storage/persistent_volumes-local.go:253
      should be able to write from pod1 and read from pod2 [It]
      test/e2e/storage/persistent_volumes-local.go:254
      Feb 11 09:43:39.647: Unexpected error:
      
          <*errors.errorString | 0xc002da26a0>: {
              s: "pod \"pod-3c1c5dcc-be5e-4979-986d-d8d0b6bafb6e\" is not Running:|
              timed out waiting for the condition",
          }
          pod "pod-3c1c5dcc-be5e-4979-986d-d8d0b6bafb6e" is not Running: 
          timed out waiting for the condition
      occurred
      test/e2e/storage/persistent_volumes-local.go:784
```

**Comment by Riaankl** [comment]
/assign @soltysh @mattfarina @kow3ns 
This is to address 3 untested Apps endpoints

File: test/e2e/apps/replica_set.go [hunk 1]
```diff
@@ -340,3 +349,72 @@ func testRSAdoptMatchingAndReleaseNotMatching(f *framework.Framework) {
...
+	framework.ExpectNoError(err, "Could not Marshal JSON for patch payload")
+
+	_, err = c.AppsV1().ReplicaSets(ns).Patch(context.TODO(), rsName, types.StrategicMergePatchType, []byte(rsScalePatchPayload), metav1.PatchOptions{}, "scale")
+	framework.ExpectNoError(err, "Failed to patch replicaset: %v", err)
+
+	ginkgo.By("verifying the replicaset Spec.Replicas was modified")
+	rs, err = c.AppsV1().ReplicaSets(ns).Get(context.TODO(), rsName, metav1.GetOptions{})
+	framework.ExpectNoError(err, "Failed to get replicaset resource: %v", err)
+	framework.ExpectEqual(*(rs.Spec.Replicas), int32(4), "replicaset should have 4 replicas")
+	e2ereplicaset.WaitForReplicaSetTargetAvailableReplicas(c, rs, 4)
```
Comment (soltysh) [inline test/e2e/apps/replica_set.go:419]: Similarly to the other PR, we don't care about the result. This test is only ensuring the patch operation worked as expected, there's a separate test exercising scaling capabilities.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: test/e2e/apps/replica_set.go (hunk 1 above)
Comment (Riaankl) [reply test/e2e/apps/replica_set.go]: Updated as requested and rebased
Thank you for the review

File: test/e2e/apps/replica_set.go [hunk 2]
```diff
@@ -21,14 +21,19 @@ import (
 	"fmt"
 	"time"
 
+	"encoding/json"
```
[the remark this answers is not in the document]
Comment (Riaankl) [reply test/e2e/apps/replica_set.go:24]: Thank you, now I understand the different blocks. Never noticed it before.
Updated

File: test/e2e/apps/replica_set.go [hunk 3]
```diff
@@ -340,3 +349,72 @@ func testRSAdoptMatchingAndReleaseNotMatching(f *framework.Framework) {
 	})
 	framework.ExpectNoError(err)
 }
+
+func testRSScaleSubresources(f *framework.Framework) {
+
```
[the remark this answers is not in the document]
Comment (Riaankl) [reply test/e2e/apps/replica_set.go:354]: Line removed

File: test/e2e/apps/replica_set.go [hunk 4]
```diff
@@ -340,3 +349,72 @@ func testRSAdoptMatchingAndReleaseNotMatching(f *framework.Framework) {
...
+	// Create webserver pods.
+	rsPodLabels := map[string]string{
+		"name": "sample-pod",
+		"pod":  WebserverImageName,
+	}
+
+	rsName := "test-rs"
+	replicas := int32(1)
+	rsRevision := "3546343826724305832"
```
[the remark this answers is not in the document]
Comment (Riaankl) [reply test/e2e/apps/replica_set.go:366]: rsRevision removed

File: test/e2e/apps/replica_set.go [hunk 5]
```diff
@@ -340,3 +349,72 @@ func testRSAdoptMatchingAndReleaseNotMatching(f *framework.Framework) {
...
+	err = e2epod.VerifyPodsRunning(c, ns, "sample-pod", false, replicas)
+	framework.ExpectNoError(err, "error in waiting for pods to come up: %s", err)
+
+	ginkgo.By("getting scale subresource")
+	scale, err := c.AppsV1().ReplicaSets(ns).GetScale(context.TODO(), rsName, metav1.GetOptions{})
+	if err != nil {
+		framework.Failf("Failed to get scale subresource: %v", err)
+	}
+	framework.ExpectEqual(scale.Spec.Replicas, int32(1), "")
+	framework.ExpectEqual(scale.Status.Replicas, int32(1), "")
```
[the remark this answers is not in the document]
Comment (Riaankl) [reply test/e2e/apps/replica_set.go]: Removed the empty strings

File: test/e2e/apps/replica_set.go [hunk 6]
```diff
@@ -340,3 +349,72 @@ func testRSAdoptMatchingAndReleaseNotMatching(f *framework.Framework) {
...
+	rsPodLabels := map[string]string{
+		"name": "sample-pod",
+		"pod":  WebserverImageName,
+	}
+
+	rsName := "test-rs"
+	replicas := int32(1)
+	rsRevision := "3546343826724305832"
+	annotations := make(map[string]string)
+	annotations[deploymentutil.RevisionAnnotation] = rsRevision
```
[the remark this answers is not in the document]
Comment (Riaankl) [reply test/e2e/apps/replica_set.go:368]: Updated and simplified. Good point. Thanks you.

File: test/e2e/apps/replica_set.go [hunk 7]
```diff
@@ -340,3 +348,63 @@ func testRSAdoptMatchingAndReleaseNotMatching(f *framework.Framework) {
...
+	// Create webserver pods.
+	rsPodLabels := map[string]string{
+		"name": "sample-pod",
+		"pod":  WebserverImageName,
+	}
+
+	rsName := "test-rs"
+	replicas := int32(1)
+	ginkgo.By(fmt.Sprintf("Creating replica set %q that asks for more than the allowed pod quota", rsName))
+	rs := newRS(rsName, 1, rsPodLabels, WebserverImageName, WebserverImage, nil)
```
Comment (dims) [inline test/e2e/apps/replica_set.go:365]: 2nd parameter is `replicas` right?

Comment (dims) [reply test/e2e/apps/replica_set.go:365]: let's use the variable instead of the hard coded `1`

Comment (Riaankl) [reply test/e2e/apps/replica_set.go:365]: Yes it is.

File: test/e2e/apps/replica_set.go (hunk 1 above)
Comment (dims) [reply test/e2e/apps/replica_set.go]: @soltysh are you saying that we should keep the Get and verify `rs.Spec.Replicas` is 4 ... But drop the `WaitForReplicaSetTargetAvailableReplicas`? looks like latest commit does not have the former.

File: test/e2e/apps/replica_set.go [hunk 8]
```diff
@@ -340,3 +348,63 @@ func testRSAdoptMatchingAndReleaseNotMatching(f *framework.Framework) {
...
+	rsScalePatchPayload, err := json.Marshal(autoscalingv1.Scale{
+		Spec: autoscalingv1.ScaleSpec{
+			Replicas: scale.Spec.Replicas,
+		},
+	})
+	framework.ExpectNoError(err, "Could not Marshal JSON for patch payload")
+
+	_, err = c.AppsV1().ReplicaSets(ns).Patch(context.TODO(), rsName, types.StrategicMergePatchType, []byte(rsScalePatchPayload), metav1.PatchOptions{}, "scale")
+	framework.ExpectNoError(err, "Failed to patch replicaset: %v", err)
+
```
Comment (dims) [inline test/e2e/apps/replica_set.go:409]: i know we don't want to wait for the operation to complete, but left a question to @soltysh in the previous thread asking if we still need to do the `Get` and verify if the patch succeeded. ( i know we are checking if the `Patch` operation returns an error or not, but if should we confirm that with a Get )

File: test/e2e/apps/replica_set.go (hunk 7 above)
Comment (Riaankl) [reply test/e2e/apps/replica_set.go]: Yes, I should have used the var. Updated 
Thank you.

File: test/e2e/apps/replica_set.go (hunk 8 above)
Comment (Riaankl) [reply test/e2e/apps/replica_set.go:409]: You are correct @Dims I should only have removed the last line.
Replace the `get` check.
Thank you.

Comment (dims) [reply test/e2e/apps/replica_set.go:409]: 👍🏾

**Review by dims — APPROVED** [review]
LGTM

**Comment by Riaankl** [comment]
/test pull-kubernetes-e2e-kind

Unrelated flake
```
Kubernetes e2e suite: [sig-storage] PersistentVolumes-local [Volume type: dir-link-bindmounted] One pod requesting one prebound PVC should be able to mount volume and write from pod1 expand_less | 2m19s
-- | --
test/e2e/storage/persistent_volumes-local.go:214 Feb 11 21:24:40.686: Unexpected error:     <*errors.errorString \| 0xc00055e820>: {         s: "pod \"pod-1209164f-dd11-4868-9386-18091d3dfd3e\" is not Running: timed out waiting for the condition",     }     pod "pod-1209164f-dd11-4868-9386-18091d3dfd3e" is not Running: timed out waiting for the condition occurred test/e2e/storage/persistent_volumes-local.go:217
```

**Review by dims — APPROVED** [review]
/lgtm

File: test/e2e/apps/replica_set.go (hunk 1 above)
Comment (soltysh) [reply test/e2e/apps/replica_set.go]: @dims that is correct, the test itself is ensuring that update/patch operation on scale resource succeeded. We have a separate test ensuring that scale operation works as a whole.

File: test/e2e/apps/replica_set.go [hunk 9]
```diff
@@ -18,10 +18,14 @@ package apps
...
 import (
 	"context"
+	"encoding/json"
 	"fmt"
 	"time"
 
 	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	"k8s.io/apimachinery/pkg/types"
+
```
Comment (soltysh) [inline test/e2e/apps/replica_set.go:28]: Nit: unnecessary empty line.

**Review by soltysh — APPROVED** [review]
/approve

**Comment by Riaankl** [comment]
/test pull-kubernetes-e2e-kind

Unrelated flake
```

1/5721 Tests Failed. | expand_less
-- | --
Kubernetes e2e suite: [sig-network] Proxy version v1 A set of valid responses are returned for both pod and service ProxyWithPath [Conformance] expand_less1m1stest/e2e/framework/framework.go:640 Feb 12 13:41:08.903: Pod didn't start within time out period Unexpected error:     <*errors.errorString \| 0xc0002761d0>: {         s: "timed out waiting for the condition",     }     timed out waiting for the condition occurred test/e2e/network/proxy.go:313 | Kubernetes e2e suite: [sig-network] Proxy version v1 A set of valid responses are returned for both pod and service ProxyWithPath [Conformance] expand_less | 1m1s | test/e2e/framework/framework.go:640 Feb 12 13:41:08.903: Pod didn't start within time out period Unexpected error:     <*errors.errorString \| 0xc0002761d0>: {         s: "timed out waiting for the condition",     }     timed out waiting for the condition occurred test/e2e/network/proxy.go:313
Kubernetes e2e suite: [sig-network] Proxy version v1 A set of valid responses are returned for both pod and service ProxyWithPath [Conformance] expand_less | 1m1s
test/e2e/framework/framework.go:640 Feb 12 13:41:08.903: Pod didn't start within time out period Unexpected error:     <*errors.errorString \| 0xc0002761d0>: {         s: "timed out waiting for the condition",     }     timed out waiting for the condition occurred test/e2e/network/proxy.go:313

```

**Pull request merged**